OpenDreamServer/Dream/Procs/DMProc.cs

```diff
@@ -0,0 +1,281 @@
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Text;
+using OpenDreamServer.Dream.Objects;
+using OpenDreamShared.Dream.Procs;
+
+namespace OpenDreamServer.Dream.Procs {
+    class DMProc : DreamProc {
+        public byte[] Bytecode { get; }
+
+        public DMProc(string name, DreamProc superProc, List<String> argumentNames, List<DMValueType> argumentTypes, byte[] bytecode)
+            : base(name, superProc, argumentNames, argumentTypes)
+        {
+            Bytecode = bytecode;
+        }
+
+        public override DMProcState CreateState(ExecutionContext context, DreamObject src, DreamObject usr, DreamProcArguments arguments)
+        {
+            return new DMProcState(this, context, src, usr, arguments);
+        }
+    }
+
+    class DMProcState : ProcState
+    {
+        delegate ProcStatus? OpcodeHandler(DMProcState state);
+
+        // TODO: This pool is not returned to if the proc runtimes
+        private static ArrayPool<DreamValue> _dreamValuePool = ArrayPool<DreamValue>.Shared;
+
+        #region Opcode Handlers
+        //In the same order as the DreamProcOpcode enum
+        private static OpcodeHandler[] _opcodeHandlers = new OpcodeHandler[] {

```

Comment:
```suggestion
        private static readonly OpcodeHandler[] _opcodeHandlers = {
```
can be marked readonly - immutable
also avoid object allocation and use array initialization instead